### ISSUE TITLE
机器学习-激励相关(2)

### ISSUE THREAD

**Issue by ws456999** [ws456999/tensorflow-learning-record#2]
## 激励函数
* gradient descent
* conjugate gradient
* BFGS
* L-BFGS

更加先进的激励函数，就不用自己选择学习速率，他会自己在内部尝试不同的rl，然后选择一个合适的，这样就不用自己定义

## multiclass classification
![](http://ok7py3vze.bkt.clouddn.com/C92A9957-8DFF-415A-B693-C6BA955E3481.png)

如果有很多个分类的话，可以先单独拿出一个类，然后跟其他数据做classify
如上图所示，这样我们就有了三个分类器
然后输入数据的话，其实就是讲数据输入到三个分类器中，取输出值max的那个

## 正则化
过度拟合的问题，将会在过多变量的时候出现
![](http://ok7py3vze.bkt.clouddn.com/59F47872-4988-427A-8F24-A1BF9CD66E62.png)
比如这样的情况

有两种方法可以解决问题
1. 
* 减少变量（featrue）ps. 模型选择算法
* 人工选择feature

2. 正则化
3. 保留所有特征变量，theta(J)  其中每一个变量都会对结果产生一点影响，每个数值的大小都是不同的，权重不同

## 正则化 cost function
regularization

![](http://ok7py3vze.bkt.clouddn.com/7A98D79A-308C-4A29-B55A-5F759DE9B7DE.png)

为什么选择惩罚某些featrue，这个选择的原理是什么？
是不是根据cost function

如果每个参数值都比较小的话，那么我们会得到一个更大简单的假设函数，
也就是

![](http://ok7py3vze.bkt.clouddn.com/5C7CC385-4229-43E8-8A03-253FA204553C.png)
主要就是讲cost 函数后面加上 系数 * theta(j)
这样可以防止过拟合因为每个系数都比较小

## 线性回归加上正则化
![](http://ok7py3vze.bkt.clouddn.com/3E246AD1-D44A-4903-872D-7E61FDAC3FD5.png)
这步就是将thetaj 乘以一个比1小一点点的数字，这样对结果没有什么影响，但是如果这个值太大的话， 那么就要惩罚这个theta

![](http://ok7py3vze.bkt.clouddn.com/AFB5266A-3D48-4573-9598-E356AD2F83B0.png)
标准化求解
![](http://ok7py3vze.bkt.clouddn.com/76A0DE0B-DE3B-44CE-8460-ACEEE575A070.png)

## 逻辑回归 + 正则化
[机器学习（Machine Learning）- 吴恩达（Andrew Ng）_哔哩哔哩 (゜-゜)つロ 干杯~-bilibili](https://www.bilibili.com/video/av9912938/?p=43)
完全没看懂

![](http://ok7py3vze.bkt.clouddn.com/976FFFBD-E187-4372-99C2-B1EA191066E1.png)

这张图对理解后面的东西非常重要：
1. 首先梯度下降其实就是不断的修正theta
2.  偏导(激励函数(假定函数 )- y)，求出cost function 在某个点的偏导，就是确定theta的移动方向
3. 正则化的话，就是在偏导内的函数 + (入/m) theta j 这样就可以防止过拟合

theta 0 不需要加  (入/m) theta j 这个就很厉害

![](http://ok7py3vze.bkt.clouddn.com/2319F622-48AB-4806-8A8D-3B305F121866.png)
jVal 是什么鬼，为什么他不需要偏导，（因为它是一个常量？他不涉及到x变化？我猜是这样的）
gradient(i), 就是正常的求解w向量
